corehq/apps/cloudcare/static/cloudcare/js/debugger/debugger.js

```diff
@@ -0,0 +1,407 @@
+                self.result(response.output);
+                self.success(response.status === "accepted");
+            });
+            window.analytics.workflow('[app-preview] User evaluated XPath');
+        };
+
+        self.afterRender = function() {
+            var options = {
+                mode: 'xml',
+                viewportMargin: Infinity,
+                readOnly: true,
+            };
+            self.codeMirrorResult = CodeMirror.fromTextArea($('#evaluate-result')[0], options);
+        };
+
+        self.result.subscribe(function(newResult) {
+            self.codeMirrorResult.setValue(newResult);
+        });
+
+        self.isSuccess = function(query) {
+            return query.status === 'accepted';
+        };
+
+        self.onMouseUp = function() {
+            var text = window.getSelection().toString();
+            self.selectedXPath(text);
+        };
+
+        self.matcher = function(flag, subtext) {
+            var match, regexp, currentQuery;
+            // Match text that starts with the flag and then looks like a path.
+            regexp = new RegExp('([\\s\(]+|^)' + RegExp.escape(flag) + '([\\w/-]*)$', 'gi');
+            match = regexp.exec(subtext);
+            if (!match) {
+                return null;
+            }
+            currentQuery = match[2];
+            if (currentQuery.length < 2) {
+                return null;
+            }
+            return currentQuery;
+        };
+
+        /**
+         * Set autocomplete for xpath input.
+         *
+         * @param {Array} autocompleteData - List of questions to be autocompleted for the xpath input
+         */
+        self.autocomplete = function(autocompleteData) {
+            self.$xpath = $('#xpath');
+            self.$xpath.atwho('destroy');
+            self.$xpath.atwho('setIframe', window.frameElement, true);
+            self.$xpath.off('inserted.atwho');
+            self.$xpath.on('inserted.atwho', function(atwhoEvent, $li) {
+                var input = atwhoEvent.currentTarget;
+
+                // Move cursor back one so we are inbetween the parenthesis
+                if (input.setSelectionRange && $li.data().itemData.type === 'Function') {
+                    input.setSelectionRange(input.selectionStart - 1, input.selectionStart - 1);
+                }
+            });
+            self.$xpath.atwho({
+                at: '',
+                suffix: '',
+                data: autocompleteData,
+                searchKey: 'value',
+                maxLen: Infinity,
+                highlightFirst: false,
+                displayTpl: function(d) {
+                    var icon = getIconFromType(d.type);
+                    return '<li><i class="' + icon + '"></i> ${value}</li>';
+                },
+                insertTpl: '${value}',
+                callbacks: {
+                    matcher: self.matcher,
+                },
+            });
+        };
+    };
+
+    var getIconFromType = function(type) {
```

Comment:
Stylistic, but this would be a lot more concise if you just used an object like `{'Trigger': 'fcc fcc-fd-text', ...}`